prov/util/src/util_coll.c

```diff
@@ -788,61 +793,68 @@ int ofi_join_collective(struct fid_ep *ep, fi_addr_t coll_addr,
 		util_coll_cid_initialized = 1;
 	}
 
+	// set up the new mc for future collectives
 	new_coll_mc->mc_fid.fid.fclass = FI_CLASS_MC;
 	new_coll_mc->mc_fid.fid.context = context;
 	new_coll_mc->mc_fid.fid.ops = &util_coll_fi_ops;
 	new_coll_mc->mc_fid.fi_addr = (uintptr_t) new_coll_mc;
-
+	new_coll_mc->av_set = av_set;
 	new_coll_mc->ep = ep;
+
 	coll_mc->ep = ep;
-	new_coll_mc->av_set = av_set;
+
+	ret = util_coll_state_init(coll_mc, &join_state);
+	if (ret) {
+		goto err1;
+	}
 
 	/* get the rank */
 	util_coll_find_my_rank(ep, new_coll_mc);
 	util_coll_find_my_rank(ep, coll_mc);
 
-	comp_data = calloc(1, sizeof(*comp_data));
-	if (!comp_data) {
-		ret =  -FI_ENOMEM;
-		goto err1;
-	}
-
 	if (new_coll_mc->my_rank != FI_ADDR_NOTAVAIL) {
-		memcpy(comp_data->cid_buf, util_coll_cid,
-		       OFI_CONTEXT_ID_SIZE * sizeof(uint64_t));
+		ret = util_coll_sched_copy(coll_mc, util_coll_cid,
+					   &join_state->data.join.cid_buf,
+					   OFI_CONTEXT_ID_SIZE * sizeof(uint64_t),
+					   FI_UINT8, NO_BARRIER);
+		if (ret) {
+			goto err2;
+		}
 	} else {
-		util_coll_init_cid(comp_data->cid_buf);
+		util_coll_init_cid(join_state->data.join.cid_buf);
 	}
 
-	ret = util_coll_allreduce(coll_mc, comp_data->cid_buf,
-				  comp_data->tmp_cid_buf,
-				  OFI_CONTEXT_ID_SIZE, FI_INT64,
-				  FI_BAND);
-	if (ret)
+	ret = util_coll_allreduce(coll_mc, &join_state->data.join.cid_buf,
+				  &join_state->data.join.tmp_cid_buf,
+				  OFI_CONTEXT_ID_SIZE, FI_INT64, FI_BAND);
+	if (ret) {
 		goto err2;
+	}
 
-	ret = util_coll_sched_comp(coll_mc, UTIL_COLL_JOIN_OP,
-			     comp_data, util_coll_join_comp);
-	if (ret)
+	ret = util_coll_sched_comp(coll_mc, UTIL_COLL_JOIN_OP, context,
+				   join_state, util_coll_join_comp);
+	if (ret) {
 		goto err2;
+	}
```

Comment:
nit: There are several places like this, but single line if () statements do not require braces.  It helps keep more code on the screen.